### PR TITLE
deploy: Add a way to deploy arbiter/stretch cluster

### PR DESCRIPTION
**Pull request by iamniting**
With this PR we will be able to deploy an arbiter/stretch cluster.

Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

cc @raghavendra-talur

### CODE REVIEW

**Review by raghavendra-talur — APPROVED** [review]

**Comment by raghavendra-talur** [comment]
I have reviewed the code from the deployment perspective. It has the following changes:

1. Adds a yaml for arbiter specific configuration
2. Edits the default yaml for make arbiter false
3. Adds logic to create a Storage Cluster Spec that has arbiter enabled
4. Adds intelligence to automatically determine the arbiter location which is usually done by the admin using the console. It also modifies the labeling mechanism for rest of the OCS nodes.

After merging this PR:
1. The non-arbiter setups/tests should work as before.
2. Arbiter deployment itself should succeed but the tests/checks for the deployment might fail.

I believe this PR is ready for review. I guess we could get fixes for tests/checks to make them arbiter compatible in subsequent PRs.

File: ocs_ci/deployment/deployment.py
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:406]: replace "node-role.kubernetes.io/master" with constants.MASTER_LABEL

Comment (iamniting) [reply ocs_ci/deployment/deployment.py]: Done
In addition, changed the worker label also.

Comment (ebenahar) [reply ocs_ci/deployment/deployment.py]: thanks!

File: ocs_ci/deployment/deployment.py
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:397]: could `failure-domain.beta.kubernetes.io/zone` be saved as a constant?

File: ocs_ci/deployment/deployment.py
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:404]: same here

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py:397]: Let me do that

Comment (iamniting) [reply ocs_ci/deployment/deployment.py]: Done

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py]: Done

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py]: Welcome

**Review by ebenahar — DISMISSED** [review]

**Review by vavuthu — APPROVED** [review]

**Comment by ebenahar** [comment]
I triggered a deploy job with arbiter for PR validation - https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-trigger-test-pr/3643/

File: ocs_ci/deployment/deployment.py
Comment (nehaberry) [inline ocs_ci/deployment/deployment.py:194]: @iamniting we do not get this label for on-prem cluster.

IIUC, the label which gets added as part of OCS install is as following. We need to handle this

compute-0         Ready    worker   19d   v1.19.0+7070803   app.kubernetes.io/name=vault,beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,cluster.ocs.openshift.io/openshift-storage=,kubernetes.io/arch=amd64,kubernetes.io/hostname=compute-0,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhcos,**topology.rook.io/rack=rack0**

**Review by nehaberry — CHANGES_REQUESTED** [review]

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py]: Hi Neha, as discussed on the chat, labels are one of the prereqs for the arbiter cluster. So we are assuming the admin will add those labels when there are no labels present beforehand.

I have added support for the `failure-domain.beta.kubernetes.io/zone` and `topology.kubernetes.io/zone` both of them.

File: ocs_ci/deployment/deployment.py
Comment (mbukatov) [inline ocs_ci/deployment/deployment.py:402]: Can't you just ask for a node in arbiter zone?

File: conf/ocsci/deploy_arbiter.yaml
Comment (mbukatov) [inline conf/ocsci/deploy_arbiter.yaml:4]: How do I define which is the arbiter zone?

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py:402]: That can be done, In that case everything needs to be done by user/admin

File: conf/ocsci/deploy_arbiter.yaml
Comment (iamniting) [reply conf/ocsci/deploy_arbiter.yaml:4]: It will be determined automatically.

Comment (mbukatov) [reply conf/ocsci/deploy_arbiter.yaml:4]: This won't work for testing purposes.

Comment (iamniting) [reply conf/ocsci/deploy_arbiter.yaml:4]: Can you please explain why? I have tested it and it worked. BTW arbiter setups have some special requirements.

Comment (iamniting) [reply conf/ocsci/deploy_arbiter.yaml:4]: We are doing some autodetection in the UI

Comment (mbukatov) [reply conf/ocsci/deploy_arbiter.yaml:4]: - From OCS Operator perspective, there is no autodetection. StorageCluster CR have to specify all the details (as you noted, some autodetection is happening only in UI).
- We want to be in full control of the deployment process, as we do for other CI deployment configurations, which eg. specifies all the zones where we care about zones.
- Autodetection in ocs-ci would be error prone and could invalidate our testing if eg. we deploy it in a wrong environment. We would be debugging ocs-ci code instead of a product.

Comment (iamniting) [reply conf/ocsci/deploy_arbiter.yaml:5]: Hi @mbukatov I have introduced functionality to feed zones via yaml conf files. Pls take a look

**Review by raghavendra-talur — APPROVED** [review]

File: conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml
Comment (prsurve) [inline conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml:17]: we should use i3 instead of  m5.4xlarge because we will get a direct nvme disk attached to an instance

File: conf/ocsci/deploy_arbiter.yaml
Comment (prsurve) [inline conf/ocsci/deploy_arbiter.yaml:8]: i guess this need's to be generated dynamically because this will not work for VMware deploymeny

File: ocs_ci/deployment/deployment.py
Comment (prsurve) [inline ocs_ci/deployment/deployment.py:396]: We also need a function to add zone label in case of vmware deployment

File: conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml
Comment (iamniting) [reply conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml:17]: I can change this. Can you pls specify the full name.

File: conf/ocsci/deploy_arbiter.yaml
Comment (iamniting) [reply conf/ocsci/deploy_arbiter.yaml:8]: There are 2 ways to deploy an arbiter, one is static and another one is dynamic.
[1] In static, you need to feed zones in conf file `conf/ocsci/deploy_arbiter.yaml`
[2] In dynamic, zones will get autodetect `conf/ocsci/deploy_arbiter_autodetect.yaml`

Note: For both deployments, you need to use different conf files.

File: ocs_ci/deployment/deployment.py
Comment (iamniting) [reply ocs_ci/deployment/deployment.py:396]: I am expecting that should be done exactly after OCP deployment, not in b/w OCS deployments. Moreover, there should be some pattern to add zone labels.

File: conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml
Comment (Akarsha-rai) [reply conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml:17]: @iamniting i3en.2xlarge

Comment (iamniting) [reply conf/deployment/aws/ipi_3az_rhcos_3m_4w.yaml]: Done

**Comment by obnoxxx** [comment]
It seems to me the requests have been addressed and questions have been answered.
What else is needed?
Can we move forward?

@ebenahar , @RazTamir , @mbukatov , @nehaberry , @prsurve  ...

**Comment by prsurve** [comment]
@iamniting can you please rebase the code and resolve the conflicts

**Comment by iamniting** [comment]
> @iamniting can you please rebase the code and resolve the conflicts

Done

**Comment by petr-balogh** [comment]
Hey @iamniting , I am working on this PR: https://github.com/red-hat-storage/ocs-ci/pull/3892 which should allow us to test on other then I3 instance types. It suppose to deploy OCP cluster on m5.4xlarge instance type and will attach local EBS volumes. So then we can do LSO kind of deployment there without need of I3 which is only tech preview.  Testing now the OCP deployment, if it will work, then I can try to test your PR on top of this cluster with OCS deployment.   What is the current status of this PR? Is it ready to be tested?

**Comment by iamniting** [comment]
> Hey @iamniting , I am working on this PR: #3892 which should allow us to test on other then I3 instance types. It suppose to deploy OCP cluster on m5.4xlarge instance type and will attach local EBS volumes. So then we can do LSO kind of deployment there without need of I3 which is only tech preview. Testing now the OCP deployment, if it will work, then I can try to test your PR on top of this cluster with OCS deployment. What is the current status of this PR? Is it ready to be tested?

Hello @petr-balogh It is ready, I have tested it once and was able to deploy a setup on aws

**Comment by petr-balogh** [comment]
@iamniting can you please rebase, I see another conflict here.

Thanks!

**Comment by iamniting** [comment]
> @iamniting can you please rebase, I see another conflict here.
> 
> Thanks!

Done, I hope I will see it getting merged very soon.

File: ocs_ci/deployment/deployment.py
Comment (nehaberry) [reply ocs_ci/deployment/deployment.py]: ack

**Review by nehaberry — DISMISSED** [review]

**Review by petr-balogh — APPROVED** [review]

**Comment by petr-balogh** [comment]
Both deployment passed, now waiting with destroy of the clusters and cause of PSI OpenStack issue the jobs are waiting till PSI OpenStack is up.  After that the jobs will report here but already we can consider as verified.  Looked to both clusters manually and arbiter mon pod was located on master node on `us-east-2c` zone an ind storageCluster CR yaml everything looked also OK.  This PR is now ready to merge and if something else will need to be done we can follow up in other PR to not block testing of Arbiter feature.

**Review by jimdevops19 — APPROVED** [review]

**Pull request merged**
